src/types.rs

```diff
@@ -184,11 +184,19 @@ pub trait TypedNode: Clone {
 }
 
 pub trait TokenWrapper: TypedNode {
-    fn as_str(&self) -> &str {
-        match &self.node().green().children().next() {
-            Some(rowan::NodeOrToken::Token(token)) => token.text(),
-            _ => unreachable!(),
-        }
+    fn to_inner_token(&self) -> SyntaxToken {
+        self.node()
+            .children_with_tokens()
+            .filter_map(|element| element.into_token())
+            .next()
+            .unwrap()
+    }
+
+    /// ### Migration note about `as_str` (from `v0.10`)
+    /// `TokenWrapper::as_str` usage (`x.as_str()`) should be replaced with (`x.to_inner_string()`).
+    /// Note that the original usage used references, but the new API returns an owned string
```

Comment:
I'm not sure if I understand the problem you're referring to in your op :)

Reply:
simple: `to_inner_string` allocates, `as_str` didn't. But `to_inner_string` is still sometimes more convenient than using `.to_inner_token().text()`.

Reply:
* If we say that it should be replaced with something else we de facto deprecate it and should eliminate all references from our code-base.
* Also, can you point me to the dangling-references issue you referenced? To me, this sounds like a problem.

Reply:
(a) Most occurrences are just tests/examples. Yeah, we can probably do that. I would move the note to `to_inner_token` then, which would be a "sensible upgrade path" or so...
(b) Not a problem of the code, just the impossibility of pleasing the borrow checker with a direct equivalent to `as_str`. That is, this doesn't introduce any dangling references.

Reply:
>  (b) Not a problem of the code, just the impossibility of pleasing the borrow checker with a direct equivalent to as_str. That is, this doesn't introduce any dangling references.

Ah I see, I guess I misunderstood your initial comment then, sorry!
Going to take a look, but the rest seems good :)